SI-4859 Don't rewrite CC().CC2() to new CC2

retronym opened the pull request:
Where CC and CC2 are case classes. Attempting to do so leads to
a "no legal prefix" error.

Now, we restrict this optimization (living in RefChecks ?!) to
case class applies with a "safe to inline" qualifier.

Review by @paulp
- is `isExprSafeToInline` the right check here? I wasn't sure about that.
- I didn't go as far as adding a test to show that some rewriting still works. But via manual inspection: here's how the enclosed test case looks after `refchecks`.

``` scala
  object O extends scala.AnyRef {
    def <init>(): O.type = {
      O.super.<init>();
      ()
    };
    new C().CC.apply();
    new D().DD.apply()
  };
```

_Update_

I also had to modify `Flatten` and `GenICode` to make sure we throw in:

``` scala
{???; Predef}.DummyImplicit.dummyImplicit
```

and `DeadCodeElimination` to initialize object `Outer` in `Outer.Inner.foo`.

retronym committed:
SI-4859 Don't rewrite CC().CC2() to new CC2

Where CC and CC2 are case classes. Attempting to do so leads to
a "no legal prefix" error.

Now, we restrict this optimization (living in RefChecks ?!) to
case class applies with a "safe to inline" qualifier.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1077/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1077/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1789/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1789/

paulp commented:
I'd say the test should be module.symbol.isStable. That is consistent with what is required for the constructor call. I think isExprSafeToInline would work out about the same in practice, but the reasoning behind the logic would be less apparent.

paulp commented:
Re: living in Refchecks, I proposed yesterday we split Refchecks into two phases, one which is exclusively about "checks" and one with all the other stuff. I don't think we'd be sad to have done this.

retronym commented:
Thinking about it some more, I believe that `isExprSafeToInline` is necessary. Replacing it with `.symbol.isStatic` will elide the qualifier in:

``` scala
object O {
  case class N()
}
object Test extends App {
  {???; O}.N()
}
```

But even after we change `RefChecks` to use `isExprSafeToInline`, the poor qualifier only makes it as far as `Flatten`, which discards it. It really doesn't stand a chance!

The `Flatten` problem, seen without case classes:

``` scala
scala> object O { object N { def foo = 0 } }
defined module O

scala> {???; O}.N.foo
res3: Int = 0
```

retronym commented:
> But even after we change `RefChecks` to use `isExprSafeToInline`, the poor qualifier only makes it as far as `Flatten`, which discards it. It really doesn't stand a chance!

`GenICode` also discards the qualifier, if we let him venture that far.

:musical_note: _Mister Cellophane / 'Cause you can look right through me / Walk right by me /And never know I'm there_ :musical_note:

retronym committed:
Introduce a new symbol test: isTopLevel

More intention revealing than: sym.owner.isPackageClass.

retronym committed:
SI-4859 Don't elide qualifiers when selecting nested modules.

Otherwise we fail to throw in:

   {???; Predef}.DummyImplicit.dummyImplicit

We still elide the initialization of `Outer` in `Outer.Inner.foo`
as before, although that seems a little dubious to me.

In total, we had to change RefChecks, Flatten, and GenICode
to effect this change. A recently fixed bug in tail call elimination
was also due to assuming that the the qualifier of a Select node
wasn't worthy of traversal.  Let's keep a close eye out for more
instances of this problem.

Furthermore, dead code elimination eliminates the module load
of `Outer` in `Outer.Inner.foo`, which would cause run/t4859.scala
to fail under the -optimize. We now err on the side of correctness and retain all MODULE_LOAD instructions unless a backwards
compatibility flag is set.

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1895/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1895/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1179/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1179/

paulp commented:
Wow, that is quite a find. A lot of code could conceivably depend on it working that way.

retronym commented:
I'm happy enough with `!sym.isTopLevel`.

retronym commented:
I'm a bit puzzled as to why the new test failed on Jenkins. Let's try that again...

```
  [partest] testing: [...]/files/run/t4859.scala                                  [FAILED]
  [partest] Inner
  [partest] Inner.i
  [partest] About to reference Inner.i
  [partest] Inner.i
  [partest] About to reference O.N
  [partest] About to reference O.N
  [partest] About to reference O.N.apply()
```

retronym commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1899/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1899/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1183/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1183/

retronym commented:
> I'm a bit puzzled as to why the new test failed on Jenkins. Let's try that again...

Oh, I see, the tests are re-run under `-optimize`. That's a nice surprise!

Indeed I could reproduce the test failure locally with:

```
SCALAC_OPTS="-optimize" ./test/partest test/files/run/t4859.scala
```

I've updated the patch to modify `DeadCodeElimination` to retain all `MODULE_LOAD` instructions. The prior, more aggressive behaviour is available under `-Ydead-code-module-load`.

All in all, this PR is getting a bit ambitious for `2.10.x`. I'll let the build run and garner a bit more review (@magarcia @gkossakowski) here, but I'll probably retarget the whole thing at master.

retronym commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1902/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1902/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1186/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1186/

retronym commented:
PLS REBUILD ALL (had a stray commit in the history)

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1907/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1907/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1191/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1191/

paulp commented on test/files/run/t4859.scala:
```diff
@@ -0,0 +1,29 @@
+
+object Test {
+  def main(args: Array[String]) {
+    Outer.Inner.i // we still don't initiialize Outer here (but should we?)
```
I don't see how we can fail to initialize Outer here if we wish to have at all predictable semantics.

paulp commented:
It looks good to me, except that I agree with retargeting to master.

Also maybe break isTopLevel into a separate commit, and pick off the 45 (!) uses, which may point toward further consolidation:

```
!a.symbol.owner.isPackageClass &&
!sym.owner.isPackageClass && matchesInfo
&& !s.owner.isPackageClass
&& existing.owner.isPackageClass
&& owner.isPackageClass
//println("classToJava "+clazz+" "+clazz.owner+" "+clazz.owner.isPackageClass)//debug
args.isEmpty && (sym.owner.isPackageClass || isSimple(pre))
case Select(qual, name) if (sym.isStaticModule && !sym.owner.isPackageClass) =>
case csym: ClassSymbol if csym.owner.isPackageClass => enterClassSymbol(tree, csym)
case sym if sym.owner.isPackageClass =>
def exists = !owner.isPackageClass || { rawInfo.load(this); rawInfo != NoType }
else if (!sym.owner.isPackageClass) compiles(sym.enclosingTopLevelClass)
else if (!sym.owner.isPackageClass) traverse(pre)
else if (clazz.owner.isPackageClass)
else if (owner.isPackageClass) "object "
else if (sym.owner.isPackageClass) sym.javaClassName
final def needsFlatClasses    = phase.flatClasses && rawowner != NoSymbol && !rawowner.isPackageClass
if ((clazz isSubClass ClassfileAnnotationClass) && !clazz.owner.isPackageClass)
if (m.owner.isPackageClass && !m.isPackage) {
if (name.isTermName && owner.isPackageClass) {
if (owner != nextOuter.owner && owner.isClass && !owner.isPackageClass && !inSelfSuperCall) {
if (owner.isPackageClass) owner.info.decls else op
if (owner.isPackageClass) owner.info.decls else op
if (owner.isPackageClass) owner.thisType
if (owner.isPackageClass) {
if (owner.isTerm || owner.isPackageClass) Nil
if (result.symbol.isLocal || result.symbol.owner.isPackageClass)
if (sym.owner.isPackageClass)
if (sym.owner.isPackageClass) {
if (symbol.owner.isPackageClass)
override def associatedFile = if (owner.isPackageClass) _associatedFile else super.associatedFile
override def isNestedClass           = !owner.isPackageClass
sym.isSourceMethod && sym.owner.isClass && !sym.owner.isPackageClass
val childOfTopLevel = clazz.owner.owner.isPackageClass
val inOrOut = if (prevSym.owner.isPackageClass) "outside of" else "in"
val isInner = !x.rawowner.isPackageClass
val isObject = owner.isModuleClass && !owner.isPackageClass
val pre1  = if (sym.owner.isPackageClass) sym.owner.thisType else if (qual == EmptyTree) NoPrefix else qual.tpe
val ref = if (tagModule.owner.isPackageClass) Ident(tagModule) else Select(prefix, tagModule.name)
val where = if (currentSym.owner.isPackageClass != prevSym.owner.isPackageClass) {
|| !owner.isPackageClass && owner.isLocalClass
|| isModuleOrModuleClass && (owner.isPackageClass || !settings.overrideObjects.value)
|| newS.owner.isPackageClass
} else if (owner.isPackageClass) {
} else if (scope != nextOuter.scope && !owner.isPackageClass) {
```

gkossakowski commented:
I agree with @paulp. Also, changes to inlining look good to me but if retarget against master I don't think you need a new -Y option.

retronym commented:
I'll open a new PR in a few days.
